### DISCUSSION TITLE
Input file for powerSpectrumPrimordialTransferred

### DISCUSSION THREAD

**Discussion by prakharbansal16** [galacticusorg/galacticus#551] [Q&A] [answered]
I'm trying to generate the halo mass function using an external power spectrum. For the purpose I've added the following lines to the .xml file 
```
<powerSpectrumPrimordialTransferred value ="file">
  <fileName value = "transferFunction_CDM.hdf5"/>
  </powerSpectrumPrimordialTransferred>

```
Now when I compile this I get an error "can not create/overwrite read only file 'transferFunction_CDM.hdf5'". The file 'transferFunction_CDM.hdf5' has been generated using the power spectrum data generated independently. I'm not sure what's going wrong. Any inputs would be very helpful

**Comment by abensonca** [comment, marked as answer]
Hi Prakhar. This error message most likely indicates that Galacticus can not find the file `'transferFunction_CDM.hdf5`. Check that it exists and the file name in your parameter file exactly matches that of the actual file. Also, if the file is not in the same directory path as you're running Galacticus from you'll need to specify the full path to it, e.g.:
```
<powerSpectrumPrimordialTransferred value ="file">
  <fileName value="/path/to/my/file/transferFunction_CDM.hdf5"/>
</powerSpectrumPrimordialTransferred>
```
If the file _does_ exist and the path is correct and you still get this message, the next most likely thing is that the file does not have read permissions set (that seems unlikely though), which you could fix with:
```
chmod u+r transferFunction_CDM.hdf5
```

**Reply by prakharbansal16** [reply]
Thanks for your suggestion. There was actually a path mismatch with the file. I'm running into  a new error now where it says that the version number (the file format attribute) is not correct. I've tried with the version numbers 5 and 6 (as per suggestions from google) but both don't seem to work. Could you please suggest what is the compatible version number?

**Reply by abensonca** [reply]
It needs the version number to be `1`.

**Reply by prakharbansal16** [reply]
I tried using it but it throws an error 'attribute must be a double scalar'. Then I tried using 1.0 and then i got the incorrect version number message again. I'm not sure what exactly to use here

**Reply by abensonca** [reply]
Can you share the `transferFunction_CDM.hdf5` file, or the file you creating it with, and your Galacticus parameter file? I can then look into the problem. You can e-mail them direct to me at abenson@carnegiescience.edu

**Reply by prakharbansal16** [reply]
I've just sent you a mail. Thanks for looking into my code!

**Reply by abensonca** [reply]
That looks OK to me. I would change the line:
```
    f.attrs["fileFormat"] = 0  # Adjust version number if needed
```
to
```
    f.attrs["fileFormat"] = 1  # Adjust version number if needed
```
to get the right version number.

Can you do:
```
h5dump -A transferFunction_CDM.hdf5
```
and paste the output here?

**Reply by prakharbansal16** [reply]
This is the output I get 


HDF5 "transferFunction_CDM.hdf5" {
GROUP "/" {
   ATTRIBUTE "description" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "Transfer Function data generated using CLASS for CDM"
      }
   }
   ATTRIBUTE "fileFormat" {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SCALAR
      DATA {
      (0): 1
      }
   }
   ATTRIBUTE "redshift" {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SCALAR
      DATA {
      (0): 0
      }
   }
   GROUP "darkMatter" {
      DATASET "transferFunctionZ0.0000" {
         DATATYPE  H5T_IEEE_F64LE
         DATASPACE  SIMPLE { ( 139 ) / ( 139 ) }
      }
   }
   GROUP "extrapolation" {
      GROUP "wavenumber" {
         ATTRIBUTE "high" {
            DATATYPE  H5T_STRING {
               STRSIZE H5T_VARIABLE;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "524.2809033602"
            }
         }
         ATTRIBUTE "low" {
            DATATYPE  H5T_STRING {
               STRSIZE H5T_VARIABLE;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "1.045004358465e-05"
            }
         }
      }
   }
   GROUP "parameters" {
      ATTRIBUTE "HubbleConstant" {
         DATATYPE  H5T_STRING {
            STRSIZE H5T_VARIABLE;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_UTF8;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SCALAR
         DATA {
         (0): "67.556"
         }
      }
      ATTRIBUTE "OmegaBaryon" {
         DATATYPE  H5T_STRING {
            STRSIZE H5T_VARIABLE;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_UTF8;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SCALAR
         DATA {
         (0): "0.0488"
         }
      }
      ATTRIBUTE "OmegaDarkEnergy" {
         DATATYPE  H5T_STRING {
            STRSIZE H5T_VARIABLE;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_UTF8;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SCALAR
         DATA {
         (0): "0.6874"
         }
      }
      ATTRIBUTE "OmegaMatter" {
         DATATYPE  H5T_STRING {
            STRSIZE H5T_VARIABLE;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_UTF8;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SCALAR
         DATA {
         (0): "0.3126"
         }
      }
   }
   DATASET "wavenumber" {
      DATATYPE  H5T_IEEE_F64LE
      DATASPACE  SIMPLE { ( 139 ) / ( 139 ) }
   }
}
}

**Reply by abensonca** [reply]
That all looks ok as far as I can tell. So, possibly there's a bug in Galacticus that prevents it from reading this. Can you send me the `transferFunction_CDM.hdf5` file and I'll try running it myself.

**Reply by prakharbansal16** [reply]
Sure, just sent

**Reply by abensonca** [reply]
I think I see the problem. In your Python code:
```
    parameters_group.attrs["HubbleConstant"] = "67.556"  # Replace with your value
    parameters_group.attrs["OmegaBaryon"] = "0.0488"  # Replace with your value
    parameters_group.attrs["OmegaDarkEnergy"] = "0.6874"  # Replace with your value
    parameters_group.attrs["OmegaMatter"] = "0.3126"  # Replace with your value
```
these values are strings because they're enclosed in quotes `""`. So, they are stored to the HDF5 file as strings, not as numbers, which is what Galacticus expects. If you remove the `""` around these numbers I think it should resolve this problem.

Also, for the redshift, I would do:
```
    f.attrs["redshift"] = 0.0  # Adjust redshift value as needed
```
to be sure that it is interpreted as a real number and not an integer.

**Reply by prakharbansal16** [reply]
Thanks, that was the error. However I'm running into  a new error now where it shows the attribute 'extrapolationWavenumber' is missing from the file. But in the documentation it's not mentioned to include it anywhere. What exactly should be given in this attribute field?

**Reply by abensonca** [reply]
You're right - the documentation needs to be updated!

You'll need to set to attributes:
1. `extrapolationWavenumber`;
2. `extrapolationRedshift`.

These control how the power spectrum is extrapolated beyond the tabulated ranges in wavenumber and redshift. These should be string types, and can be one of:
* `extrapolate`
* `fix`
* `abort`
* `zero`
* `powerLaw`

Often the safest is to choose `abort` if you're not sure if your tabulated results will be stable if extrapolated. Otherwise, 'extrapolate` or `powerLaw` are good options.

**Reply by prakharbansal16** [reply]
It gives an error message 'attribute must be an character scalar' when I used 'abort' for both attributes

**Reply by abensonca** [reply]
Can you do the `h5dump` command again so we can see the types of those attributes?

**Comment by prakharbansal16** [comment]
Here's the output 
HDF5 "parameters/tutorials/transferFunction_CDM.hdf5" {
GROUP "/" {
   ATTRIBUTE "description" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "Transfer Function data generated using CLASS for CDM"
      }
   }
   ATTRIBUTE "extrapolationRedshift" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "abort"
      }
   }
   ATTRIBUTE "extrapolationWavenumber" {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_UTF8;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): "abort"
      }
   }
   ATTRIBUTE "fileFormat" {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SCALAR
      DATA {
      (0): 1
      }
   }
   ATTRIBUTE "redshift" {
      DATATYPE  H5T_IEEE_F64LE
      DATASPACE  SCALAR
      DATA {
      (0): 0
      }
   }
   GROUP "darkMatter" {
      DATASET "transferFunctionZ0.0000" {
         DATATYPE  H5T_IEEE_F64LE
         DATASPACE  SIMPLE { ( 139 ) / ( 139 ) }
      }
   }
   GROUP "extrapolation" {
      GROUP "Wavenumber" {
         ATTRIBUTE "high" {
            DATATYPE  H5T_STRING {
               STRSIZE H5T_VARIABLE;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "524.2809033602"
            }
         }
         ATTRIBUTE "low" {
            DATATYPE  H5T_STRING {
               STRSIZE H5T_VARIABLE;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SCALAR
            DATA {
            (0): "1.045004358465e-05"
            }
         }
      }
   }
   GROUP "parameters" {
      ATTRIBUTE "HubbleConstant" {
         DATATYPE  H5T_IEEE_F64LE
         DATASPACE  SCALAR
         DATA {
         (0): 67.556
         }
      }
      ATTRIBUTE "OmegaBaryon" {
         DATATYPE  H5T_IEEE_F64LE
         DATASPACE  SCALAR
         DATA {
         (0): 0.0488
         }
      }
      ATTRIBUTE "OmegaDarkEnergy" {
         DATATYPE  H5T_IEEE_F64LE
         DATASPACE  SCALAR
         DATA {
         (0): 0.6874
         }
      }
      ATTRIBUTE "OmegaMatter" {
         DATATYPE  H5T_IEEE_F64LE
         DATASPACE  SCALAR
         DATA {
         (0): 0.3126
         }
      }
      ATTRIBUTE "temperatureCMB" {
         DATATYPE  H5T_IEEE_F64LE
         DATASPACE  SCALAR
         DATA {
         (0): 2.7255
         }
      }
   }
   DATASET "wavenumber" {
      DATATYPE  H5T_IEEE_F64LE
      DATASPACE  SIMPLE { ( 139 ) / ( 139 ) }
   }
}
}

**Reply by abensonca** [reply]
Thanks. I see the problem - `h5py` is storing these attributes as variable-length strings, which Galacticus currently doesn't support. However, I think it should be easy to add support for those, so I'll implement that tomorrow and let you know once it's available.

**Reply by prakharbansal16** [reply]
Sure, that would be helpful. Thank you so much!

**Reply by abensonca** [reply]
This should be fixed now - so it should be able to correctly read the attributes from your file. If you're compiling Galacticus from source then you can just update and recompile. If you're using one of the pre-compiled binaries you'll need to wait for these to be regenerated. You can watch the test/build/deploy workflow [here](https://github.com/galacticusorg/galacticus/actions/runs/7822645298) - once the spinning circle in the top left changes to a green check mark, the new version should be ready for use.

**Reply by prakharbansal16** [reply]
Thanks the error got fixed. I'm getting a new error this time "dataset 'power' has incorrect rank". I think power is an array of rank 2 but I don't understand why are 2 columns needed for that. Wanted to confirm are the formats for the two files "transferFunction'' and 'powerSpectrumPrimordialTransferred' different? I'm using the format for the 'transferFunction' as given in the Galaticus_Physics document for the 'powerSpectrumPrimordialTransferred' and so no 'power' dataset was mentioned there. If possible could you please provide the updated documentation for both these functions and the structure of the 'hdf' file that should be used as input

**Reply by abensonca** [reply]
`powerSpectrumPrimordialTransferred` does require a rank-2 array for the `power` dataset, with the two dimensions corresponding to wavenumber and redshift. This is intended for cases where you want to specify the full late-time growth of the power spectrum.

In many cases (e.g. CDM), the late-time growth is described by a wavenumber-independent linear growth factor. In that case, it would be a better option to use the `transferFunction` class with the `file` option to read a transfer function from file. Then you only need to specify this as a function of wavelength.

The documentation for the `powerSpectrumPrimordialTransferred` `file` option can be found in [Galacticus_Development.pdf](https://github.com/galacticusorg/galacticus/releases/download/bleeding-edge/Galacticus_Development.pdf) - section 3.173.1 on page 410, but is currently very limited. I'll update it now.

**Reply by prakharbansal16** [reply]
Oh okay I thought the wavenumber should be provided as a separate array. Also I'm now getting an error 'data is not interpolatable'. Could you suggest any possible fix for that?

I can't find section 3.173.1 on pg 410. The maximum section number in chapter 3 is 3.6. There is some information given in section 2.9.134 but as you mentioned that doesn't describe the file contents.

**Reply by abensonca** [reply]
You do need to supply a separate array of wavenumbers (and of redshifts). But the `power` dataset is a function of both of those, so must be rank-2.

The `data is not interpolatable` indicates that one of the arrays `wavenumber` or `redshift` has only one entry - at least 2 are needed to do interpolation and extrapolation.

My  mistake, I linked the wrong file. It should be [Galacticus_Physicst.pdf](https://github.com/galacticusorg/galacticus/releases/download/bleeding-edge/Galacticus_Physics.pdf). The details there are very limited - I added some more but it won't appear right away.

**Reply by prakharbansal16** [reply]
Thanks, I think now my hdf file format is correct as per the updated documentation as the message "M: -> Begin task: power spectrum" is displayed but still I'm running into some problem. I'm getting the error message "Galacticus experienced a floating point exception - will try to flush data before exiting."

**Reply by abensonca** [reply]
Can you post the full output here? There should be additional information that will indicate where in the code this error is occurring.

**Reply by prakharbansal16** [reply]
Galacticus experienced a floating point exception - will try to flush data before exiting.
  => Error occurred in master thread
  => The following warnings were issued:
 Using default value for parameter '[verbosityLevel]'
 Using default value for parameter '[task/wavenumberMinimum]'
 Using default class for parameter '[darkMatterParticle]'
 Using default class for parameter '[powerSpectrumWindowFunction]'
 Using default value for parameter '[cosmologicalMassVariance/toleranceTopHat]'
 Using default class for parameter '[powerSpectrum]'
#0  0x810ded in signal_handler_sigfpe
	at ./work/build/error.p.F90:512
#1  0x2ad1c7f in ???
#2  0x1118231 in __numerical_interpolation_MOD_interpolatorlinearweights
	at ./work/build/numerical.interpolation.p.F90:1155
#3  0x1f4efba in __power_spectra_primordial_transferred_MOD_filepower
	at ./work/build/structure_formation.power_spectrum.primordial.transferred.file.p.F90:367
#4  0x1f5ca9d in __power_spectra_MOD_standardpower
	at ./work/build/structure_formation.power_spectrum.standard.p.F90:296
#5  0x21235e1 in __tasks_MOD_powerspectraperform
	at ./work/build/tasks.power_spectrum.p.F90:1032
#6  0x421cd8 in galacticus
	at ./work/build/Galacticus.p.F90:141
#7  0x4056d6 in main
	at ./work/build/Galacticus.p.F90:34

Program aborted. Backtrace:
#0  0x810e00 in signal_handler_sigfpe
	at ./work/build/error.p.F90:528
#1  0x2ad1c7f in ???
#2  0x1118231 in __numerical_interpolation_MOD_interpolatorlinearweights
	at ./work/build/numerical.interpolation.p.F90:1155
#3  0x1f4efba in __power_spectra_primordial_transferred_MOD_filepower
	at ./work/build/structure_formation.power_spectrum.primordial.transferred.file.p.F90:367
#4  0x1f5ca9d in __power_spectra_MOD_standardpower
	at ./work/build/structure_formation.power_spectrum.standard.p.F90:296
#5  0x21235e1 in __tasks_MOD_powerspectraperform
	at ./work/build/tasks.power_spectrum.p.F90:1032
#6  0x421cd8 in galacticus
	at ./work/build/Galacticus.p.F90:141
#7  0x4056d6 in main
	at ./work/build/Galacticus.p.F90:34
Aborted (core dumped)

**Reply by abensonca** [reply]
This turned out to be a bug in Galacticus which was breaking the interpolation in redshift when using power spectra read from a file as you use here. This is now fixed - the source code is updated and ready to use, the pre-compiled binaries of Galacticus are rebuilding now so will be available later today.
